source/common/config/grpc_stream.h

```diff
@@ -131,11 +136,78 @@ class GrpcStream : public Grpc::AsyncStreamCallbacks<ResponseProto>,
     return false;
   }
 
+  absl::optional<Grpc::Status::GrpcStatus> getCloseStatus() { return close_status_; }
+
 private:
   void setRetryTimer() {
     retry_timer_->enableTimer(std::chrono::milliseconds(backoff_strategy_->nextBackOffMs()));
   }
 
+  // https://github.com/envoyproxy/envoy/issues/14591
+  // Log level should be reduced when the remote close failure is `Ok` or is retriable and has only
+  // been occurring for a short amount of time.
+  void logClose(Grpc::Status::GrpcStatus status, const std::string& message) {
+    if (Grpc::Status::WellKnownGrpcStatus::Ok == status) {
+      ENVOY_LOG(debug, "{} gRPC config stream closed: {}, {}", service_method_.name(), status,
+                message);
+      return;
+    }
+
+    if (!onlyWarnOnRepeatedFailure(status)) {
+      // When the failure is considered non-retriable, warn.
+      ENVOY_LOG(warn, "{} gRPC config stream closed: {}, {}", service_method_.name(), status,
+                message);
+      return;
+    }
+
+    if (!isCloseStatusSet()) {
+      // For the first failure, record its occurrence and log at the debug level.
+      ENVOY_LOG(debug, "{} gRPC config stream closed: {}, {}", service_method_.name(), status,
+                message);
+      setCloseStatus(status, message);
+      return;
+    }
+
+    uint64_t ms_since_first_close = std::chrono::duration_cast<std::chrono::milliseconds>(
+                                        time_source_.monotonicTime() - close_time_)
+                                        .count();
+    Grpc::Status::GrpcStatus close_status = close_status_.value();
+
+    if (status != close_status) {
+      // This is a different failure. Warn on both statuses and remember the new one.
+      ENVOY_LOG(warn, "{} gRPC config stream closed: {}, {} (previously {}, {} since {}ms ago)",
+                service_method_.name(), status, message, close_status, close_message_,
+                ms_since_first_close);
+      setCloseStatus(status, message);
+      return;
+    }
+
+    if (ms_since_first_close > RetryMaxDelayMs) {
+      // Warn if we are over the time limit.
+      ENVOY_LOG(warn, "{} gRPC config stream closed since {}ms ago: {}, {}", service_method_.name(),
+                ms_since_first_close, close_status, close_message_);
+      return;
+    }
+
+    // Failure is retriable and new enough to only log at the debug level.
+    ENVOY_LOG(debug, "{} gRPC config stream closed: {}, {}", service_method_.name(), status,
+              message);
+  }
+
+  bool onlyWarnOnRepeatedFailure(Grpc::Status::GrpcStatus status) {
+    return Grpc::Status::WellKnownGrpcStatus::DeadlineExceeded == status ||
+           Grpc::Status::WellKnownGrpcStatus::Unavailable == status;
```

Comment:
Can you add some comments on how you decided these?

Looking at https://grpc.github.io/grpc/core/md_doc_statuscodes.html I would naively assume `RESOURCE_EXHAUSTED` should also be included. Maybe others?

Reply:
Sure, will add a comment. I could see why `RESOURCE_EXHAUSTED` would be included, so I'll add that. At first I was thinking the resource is ambiguous and wasn't sure how likely it would be that retrying would help, but either way it seems suppressing the log for the first 30s makes sense. I have a hard time seeing how any of the others would be included though. For example, `FAILED_PRECONDITION` says
> (a) Use UNAVAILABLE if the client can retry just the failing call. (b) Use ABORTED if the client should retry at a higher level (e.g., when a client-specified test-and-set fails, indicating the client should restart a read-modify-write sequence). (c) Use FAILED_PRECONDITION if the client should not retry until the system state has been explicitly fixed.